lib/zk_recipes/cache.rb

```diff
@@ -35,30 +48,83 @@ def initialize(logger: nil, host: nil, timeout: nil, zk_opts: {})
     end
 
     def register(path, default_value, &block)
-      raise Error, "register only allowed before setup_callbacks called" unless @registerable
+      raise Error, "register only allowed before setup_callbacks called" if @zk
+      raise Error, "path already registered path=#{path.inspect}" if @registered_paths.key?(path)
 
       debug { "added path=#{path} default_value=#{default_value.inspect}" }
       @cache[path] = CachedPath.new(default_value)
-      @registered_values[path] = RegisteredPath.new(default_value, block)
+      @registered_paths[path] = RegisteredPath.new(default_value, block)
       ActiveSupport::Notifications.instrument(AS_NOTIFICATION, path: path, value: default_value)
     end
 
+    def register_runtime(path, default_value, &block)
+      raise Error, "register_runtime only allowed after zk client is created" unless @zk
+      raise Error, "path already registered path=#{path.inspect}" if registered_path?(path) || registered_runtime_path?(path)
+
+      debug { "added runtime path=#{path} default_value=#{default_value.inspect}" }
+      @runtime_cache[path] = CachedPath.new(default_value)
+      @runtime_paths[path] = RegisteredPath.new(default_value, block)
+
+      @runtime_watches[path] = @zk.register(path) do |event|
+        if event.node_event?
+          debug("node event path=#{event.path} #{event.event_name} #{event.state_name}")
+          unless update_runtime_cache(event.path)
+            @pending_runtime_updates[event.path] = nil
+            @zk.defer(method(:process_pending_runtime_updates))
+          end
+        else
+          warn("session event #{event.event_name} #{event.state_name}")
+        end
+      end
+
+      @zk.defer do
+        unless update_runtime_cache(path)
+          @pending_runtime_updates[path] = nil
+          @zk.defer(method(:process_pending_runtime_updates))
+        end
+      end
+      ActiveSupport::Notifications.instrument(AS_NOTIFICATION_RUNTIME, path: path, value: default_value)
+    end
+
+    def register_children(path, default_children = [], &block)
+      raise Error, "register_children only allowed before setup_callbacks called" if @zk
+      raise Error, "path already registered path=#{path.inspect}" if @registered_children_paths.key?(path)
+
+      debug { "added children=#{path} default_children=#{default_children.inspect}" }
+      @children_cache[path] = CachedChildrenPath.new(default_children)
+      @registered_children_paths[path] = RegisteredChildrenPath.new(default_children, block)
+      ActiveSupport::Notifications.instrument(AS_NOTIFICATION_CHILDREN, path: path, children: default_children)
+    end
+
     def setup_callbacks(zk)
-      raise Error, "setup_callbacks can only be called once" unless @registerable
+      raise Error, "setup_callbacks can only be called once" if @zk
```

Comment:
Yes, setup is *not* threadsafe.

Reply:
That sounds fine for 90% of cases. One small case I'm worried about is in the runtime paths, though. Those registrations happen in a callback from a children update; maybe *those* are single-threaded, but it seems like a complex chain of reasoning to rely on.

Reply:
Ah, so while it is ok that `setup_callbacks` is not threadsafe, the fact that registrations are not threadsafe could be a problem. Thoughts on adding a mutex around the `register*` methods? I don't think that can deadlock.

Reply:
I think I have resolved this issue. `setup_callbacks` isn't threadsafe, but it can only be called before the ZK client connects, so the zk thread won't be running in any meaningful capacity yet (no watches, etc. will do anything).

Runtime registrations can be triggered both by the callbacks setup in `setup_callbacks` and from application code. I _don't think_ there are any thread safety issues with that.

The runtime registrations use the watch as the authoritative mutex, and the runtime path is only written when the watch mutex is held.